File: BFT,DFT/graph_search.c
#include <stdio.h>
#include <stdlib.h>

#define MAX 100

struct Graph {
    int numVertices;
    int adjMatrix[MAX][MAX];
};

void initGraph(struct Graph* graph, int vertices) {
    graph->numVertices = vertices;
    for (int i = 0; i < vertices; i++) {
        for (int j = 0; j < vertices; j++) {
            graph->adjMatrix[i][j] = 0;
        }
    }
}

void addEdge(struct Graph* graph, int src, int dest) {
    graph->adjMatrix[src][dest] = 1;
    graph->adjMatrix[dest][src] = 1;
}

void _DFS(struct Graph* graph, int vertex, int visited[]) {
    visited[vertex] = 1;
    printf("%d->", vertex);

    for (int i = 0; i < graph->numVertices; i++) {
        if (graph->adjMatrix[vertex][i] == 1 && !visited[i]) {
            _DFS(graph, i, visited);
        }
    }
}

void DFS(struct Graph* graph, int startVertex) {
    int visited[MAX] = {0};
    _DFS(graph, startVertex, visited);
}


void BFS(struct Graph* graph, int startVertex) {
    int visited[MAX] = {0};
    int queue[MAX];
    int front = 0, rear = 0;

    visited[startVertex] = 1;
    queue[rear++] = startVertex;

    while (front < rear) {
        int currentVertex = queue[front++];
        printf("%d->", currentVertex);

        for (int i = 0; i < graph->numVertices; i++) {
            if (graph->adjMatrix[currentVertex][i] == 1 && !visited[i]) {
                visited[i] = 1;
                queue[rear++] = i;
            }
        }
    }
}


int main() {
    struct Graph graph;
    initGraph(&graph, 5);

    addEdge(&graph, 0, 1);
    addEdge(&graph, 0, 2);
    addEdge(&graph, 1, 3);
    addEdge(&graph, 1, 4);

    printf("BFS from 0:");
    BFS(&graph, 0);
    printf("\n");
    printf("DFS from 0:");
    DFS(&graph, 0);
    printf("\n");
    return 0;
}

File: BFT,DFT/graph_search.py
def BFS(graph, start):
    visited = set()
    queue = [start]

    while queue:
        vertex = queue.pop(0)
        if vertex not in visited:
            visited.add(vertex)
            print(vertex, end='->')
            queue.extend(neighbor for neighbor in graph[vertex] if neighbor not in visited)

def _DFS(graph, node, visited):
    visited[node] = True
    print(node, end='->')

    for neighbor in graph[node]:
        if not visited[neighbor]:
            _DFS(graph,neighbor, visited)

def DFS(graph, start):
    visited = [False for _ in range(len(graph))]
    _DFS(graph, start, visited)

if __name__ == "__main__":
    graph = {}
    graph[0] = [1,2]
    graph[1] = [0,3,4]
    graph[2] = [0]
    graph[3] = [1]
    graph[4] = [1]

    print("BFS from 0: ", end = "")
    BFS(graph, 0)
    print("")
    print("DFS from 0: ", end = "")
    DFS(graph,0)